Add course group management

madwau opened the pull request:
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [X] Server: I added `@PreAuthorize` and check the course groups for all new REST Calls (security)
- [X] Server: I implemented the changes with a good performance and prevented too many database calls
- [X] Server: I documented the Java code using JavaDoc style.
- [X] Client: I added `authorities` to all new routes and check the course groups for displaying navigation elements (links, buttons)
- [X] Client: I documented the TypeScript code using JSDoc style.
- [x] Client: I added multiple screenshots/screencasts of my UI changes
- [X] Client: I translated all the newly inserted strings into German and English

### Motivation and Context
This PR makes the course group management easier for instructors and admins. It allows them to click on the Access Groups (Students, Tutors, Instructors) of a course in the Course Administration in order to inspect them. This opens up the new course group page that displays all users that belong to the selected course group in a table layout.

Two different actions can be performed on the page:
- Users can be removed from the course group (requires confirmation in modal)
- Users can be added to the course group via the autocomplete search bar

### Description
1. Add missing server endpoints for querying users
2. Mirror new server endpoints into client services
3. Create new component `CourseGroupComponent`
  3.1 Based on generic data table `DataTableComponent` (our wrapper around ngx-datatable)
  3.2 Adds server side search for users on top
  3.3 Customizes the behavior for when a user is selected via autocomplete (selecting a non-member user will add them to the group; member and non-member differentiation via colors and icons)
  3.4 "Remove" button for each user to remove them from the course group (uses our delete dialog button with built in confirmation dialog)
4. Enhancements to `DataTableComponent` needed for `CourseGroupComponent` (but with high reuse potential)
  4.1 Expose autocomplete suggestion buttons for out of component UI customization
  4.2 Enable customizing the search and autocomplete select behavior
  4.3 Add state information for the server side search use case (loading, no results, failed, query too short)

### Steps for Testing
<!-- Please describe in detail how the reviewer can test your changes. -->

1. Log in to Artemis (as instructor or admin)
2. Navigate to Course Administration
3. Click on one of the Access Groups for a Course
4. Check that all Course Group members are loaded
5. Test removing a user from the Course Group
6. Search for a user that is already part of the group and verify that he is found
7. Search for a user that is not yet part of the group and add them to the group

### Screenshots
1. The new Course Group page can be accessed by clicking on these group names (for instructors of the respective course and for admins):
![OnPaste 20200326-013111](https://user-images.githubusercontent.com/7109225/77598698-c9886b80-6f02-11ea-8308-967c3935eeb8.png)

2. The new Course Group page showing all group members
![OnPaste 20200326-013211](https://user-images.githubusercontent.com/7109225/77598700-cab99880-6f02-11ea-8e26-2696e000152e.png)

3. Upon clicking on the "Remove" button, the modal explains what will happen and asks for confirmation
![OnPaste 20200326-013232](https://user-images.githubusercontent.com/7109225/77598701-cab99880-6f02-11ea-9cf7-aacc70b9fd08.png)

4. When searching for a user (by login or name) to add to the course group, a list of users is presented in an autocomplete. The ones with less opacity, a light green background and a group icon are already part of the course group (clicking on them is still possible but doesn't do anything apart from filtering for them). The regular options with white background, strong green hover color and "add user" icon are users that are not yet part of the course group (clicking on them will add them to the course group; this is indicated by a flashing green row background once added).
![OnPaste 20200326-013323](https://user-images.githubusercontent.com/7109225/77598702-cb522f00-6f02-11ea-8fde-5804afacd019.png)

5. When the search term is too short for a useful server-side search (since it would return either very many results or a too unspecific selection), an info badge informs the user that he needs to enter at least 3 characters (this nice little add-on is part of the data table itself and is now also available for Exercise Scores and Participations):
![OnPaste 20200326-014833](https://user-images.githubusercontent.com/7109225/77599182-35b79f00-6f04-11ea-9d4e-6acf3fa40725.png)

6. When the server search for users returned no results:
![OnPaste 20200326-013515](https://user-images.githubusercontent.com/7109225/77598705-cb522f00-6f02-11ea-9cd8-c743ddaa836b.png)

7. When the server search for users did not execute as expected (e.g. due to no internet):
![OnPaste 20200326-013545](https://user-images.githubusercontent.com/7109225/77598706-cbeac580-6f02-11ea-90a0-08dfc03e0c91.png)

krusche commented:
Awesome 👍👍👍  The screenshots look really great, especially the usability and user experience looks really nice! And super cool, how fast you have implemented this! Thank you! Will look into it in more detail later today 👍

madwau commented:
Thanks a lot for the positive feedback everyone!

> I noticed two things and i'm not sure if they were intended to work this way, but also it could be future work.
> 1.When i remove a user from one of the access groups he will be removed from all access groups
> 2.When i add a user he will be added to all access groups

@F4lka That is correct behavior under the assumption that these access groups all share the same name, i.e. are in fact the same access group. These are not typical production conditions though and are a specialty of our test setup where different courses and all three tiers (students, tutor, instructors) sometimes share the same access group, e.g. `artemis-dev`.

madwau commented:
> I get "Search failed". I don't think it's related to your implementation but more because of something related to the test users. Therefore I approve the PR but you might want to have a look at it anyway.

Interesting finding. The search fails with an `AccessDeniedException`. Unfortunately, the stacktrace on sentry doesn't provide much of a clue as to what is happening.

My hunch would be that the authority granting performed by `ArtemisInternalAuthenticationProvider` (as opposed to `JiraAuthenticationProvider`) is somehow not working correctly in conjunction with the `@Secured({ AuthoritiesConstants.INSTRUCTOR, AuthoritiesConstants.ADMIN })` annotation used on the search endpoint. I will investigate this.

krusche commented on src/main/java/de/tum/in/www1/artemis/web/rest/UserResource.java:
```diff
@@ -164,6 +166,26 @@ else if (managedUserVM.getGroups().stream().anyMatch(group -> !artemisAuthentica
+     * @return the ResponseEntity with status 200 (OK) and with body all users
+     */
+    @GetMapping("/users/search")
+    @Secured({ AuthoritiesConstants.INSTRUCTOR, AuthoritiesConstants.ADMIN })
```
what is the difference to `@PreAuthorize("hasAnyRole('INSTRUCTOR', 'ADMIN')")`?

madwau replied:
They are equivalent in meaning but `PreAuthorize` seems to be the preferred approach nowadays since it is more powerful than `Secured`.

> `@Secured` annotation is a legacy Spring Security 2 annotation that can be used to configure method security. It supports more than only role-based security, but does not support using Spring Expression Language (SpEL) to specify security constraints. It is recommended to use the `@PreAuthorize` annotation in new applications over this annotation.

Source: https://stackoverflow.com/a/57394049

madwau replied:
`UserResource` was the only resource in the code base still using the legacy `@Secured`. I refactored all usages to equivalent `PreAuthorize` annotations instead.

krusche commented on src/main/webapp/app/shared/delete-dialog/delete-dialog.component.html:
```diff
@@ -4,6 +4,7 @@ <h4 class="modal-title" [ngSwitch]="actionType">
             <span *ngSwitchCase="actionTypes.Delete" jhiTranslate="entity.delete.title">Confirm delete operation</span>
             <span *ngSwitchCase="actionTypes.Reset" jhiTranslate="entity.reset.title">Confirm reset operation</span>
             <span *ngSwitchCase="actionTypes.Cleanup" jhiTranslate="entity.cleanup.title">Confirm cleanup operation</span>
+            <span *ngSwitchCase="actionTypes.Remove" jhiTranslate="entity.remove.title">Confirm cleanup operation</span>
```
```suggestion
            <span *ngSwitchCase="actionTypes.Remove" jhiTranslate="entity.remove.title">Confirm remove operation</span>
```

madwau commented:
> > I get "Search failed". I don't think it's related to your implementation but more because of something related to the test users. Therefore I approve the PR but you might want to have a look at it anyway.
> 
> Interesting finding. The search fails with an `AccessDeniedException`. Unfortunately, the stacktrace on sentry doesn't provide much of a clue as to what is happening.
> 
> My hunch would be that the authority granting performed by `ArtemisInternalAuthenticationProvider` (as opposed to `JiraAuthenticationProvider`) is somehow not working correctly in conjunction with the `@Secured({ AuthoritiesConstants.INSTRUCTOR, AuthoritiesConstants.ADMIN })` annotation used on the search endpoint. I will investigate this.

The issue was much simpler and more logical than that: `UserResource` has the class-level security annotation `@PreAuthorize("hasAnyRole('ADMIN')")`. Now, some of the methods in the resource used the `@Secured` method-level annotation instead of `@PreAuthorize` which is a problem since those do not override the class-level `@PreAuthorize` security annotation.

Only one existing endpoint was affected, namely `GET /users/:login : get the "login" user` which is supposed to be accessible by TAs and above but due to using the `@Secured` method-level annotation, it was in fact only ever accessible to admins all this time since the class-level `@PreAuthorize("hasAnyRole('ADMIN')")` blocked the other roles out.

The same thing happened to the new endpoint `GET /users/search` from this PR. Despite the method-level clearance for instructors, instructors were still blocked due to the non-overriding of the class-level `@PreAuthorize` by the method-level `@Secured`.

All this is fixed now since I removed `@Secured` completely and replaced it by `@PreAuthorize` (more on why `@PreAuthorize` is preferred over `@Secured` in my other [comment](https://github.com/ls1intum/Artemis/pull/1281#discussion_r399674228) to Stephan).

@sascha11110 I tested it for your exact scenario on the test server and it works for me now. Would be nice if you could maybe also confirm it. Thanks again for catching this bug which turned out to be rather critical (not in a security sense but in a usage sense).

madwau commented:
> Great work! I only found some minor issues in the code. Would be cool if you could address them.
> 
> Could you also please add some basic tests (at least for the new server REST calls), so that we do not destroy this beautiful functionality in the future? Thanks!!!

I addressed your comments. Sure, I'll add some tests now.

madwau commented:
Server tests are added for all new endpoints.

krusche commented:
Great 👍  Thanks for the server tests and the additional explanations. Really very good work!!!
